core/src/test/java/io/grpc/internal/ClientCallImplTest.java

```diff
@@ -812,18 +817,52 @@ public void expiredDeadlineCancelsStream_CallOptions() {
         provider,
         deadlineCancellationExecutor,
         channelCallTracer,
-        false /* retryEnabled */);
+        /* retryEnabled= */ false);
 
     call.start(callListener, new Metadata());
 
-    fakeClock.forwardNanos(TimeUnit.SECONDS.toNanos(1) + 1);
+    fakeClock.forwardNanos(TimeUnit.SECONDS.toNanos(1));
 
+    // Verify cancel sent to application when deadline just past
+    verify(callListener).onClose(statusArgumentCaptor.capture(),
+        ArgumentMatchers.isA(Metadata.class));
+    verify(callListener, times(1))
+        .onClose(statusCaptor.capture(), metadataArgumentCaptor.capture());
+    assertThat(statusCaptor.getValue().getDescription())
+        .matches("deadline exceeded after [0-9]+ns. \\[remote_addr=127\\.0\\.0\\.1:443\\]");
+    assertThat(statusCaptor.getValue().getCode()).isEqualTo(Code.DEADLINE_EXCEEDED);
+    verify(stream, never()).cancel(statusCaptor.capture());
+
+    fakeClock.forwardNanos(DEADLINE_EXPIRATION_CANCEL_DELAY_NANOS);
+
+    // verify cancel send to server is delayed with DEADLINE_EXPIRATION_CANCEL_DELAY
     verify(stream, times(1)).cancel(statusCaptor.capture());
     assertEquals(Status.Code.DEADLINE_EXCEEDED, statusCaptor.getValue().getCode());
     assertThat(statusCaptor.getValue().getDescription())
         .matches("deadline exceeded after [0-9]+ns. \\[remote_addr=127\\.0\\.0\\.1:443\\]");
   }
 
+  @Test
+  public void expiredDeadlineWithoutExtraWaitingDoNotReceiveCancel() {
```

Comment:
What's covered here could be covered in `expiredDeadlineCancelsStream_CallOptions()`, where you forward the clock by `DEADLINE_EXPIRATION_CANCEL_DELAY_NANOS - 1` and verify that the stream is not cancelled, then forward by 1 and verify that the stream is now cancelled.